cmd/clusterctl/pkg/client/cluster/upgrader_info.go

```diff
@@ -0,0 +1,185 @@
+}
+
+func newUpgradeInfo(metadata *clusterctlv1.Metadata, currentVersion *version.Version, nextVersions []version.Version) *upgradeInfo {
+	// Sorts release series; this ensures also an implicit ordering of ClusterAPI/API version.
+	sort.Slice(metadata.ReleaseSeries, func(i, j int) bool {
+		return metadata.ReleaseSeries[i].Major < metadata.ReleaseSeries[j].Major ||
+			(metadata.ReleaseSeries[i].Major == metadata.ReleaseSeries[j].Major && metadata.ReleaseSeries[i].Minor < metadata.ReleaseSeries[j].Minor)
+	})
+
+	// Sorts nextVersions.
+	sort.Slice(nextVersions, func(i, j int) bool {
+		return nextVersions[i].LessThan(&nextVersions[j])
+	})
+
+	return &upgradeInfo{
+		metadata:       metadata,
+		currentVersion: currentVersion,
+		nextVersions:   nextVersions,
+	}
+}
+
+// getAPIVersionsForUpgrade return the list of ClusterAPI/API version available for a provider upgrade. e.g.
+// - If the current version of the provider support v1alpha3 ClusterAPI/APIVersion (the latest), it returns v1alpha3
+// - If the current version of the provider support v1alpha3 ClusterAPI/APIVersion but there is also the v1alpha4 ClusterAPI/APIVersion available, it returns v1alpha3, v1alpha4
+func (i *upgradeInfo) getAPIVersionsForUpgrade() []string {
+	apiVersionsForUpgrade := sets.NewString()
+	for _, releaseSeries := range i.metadata.ReleaseSeries {
+		// Drop the release series if older than the current version, because not relevant for upgrade.
+		if i.currentVersion.Major() > releaseSeries.Major || (i.currentVersion.Major() == releaseSeries.Major && i.currentVersion.Minor() > releaseSeries.Minor) {
+			continue
+		}
+		apiVersionsForUpgrade.Insert(releaseSeries.ClusterAPIVersion)
+	}
+
+	return apiVersionsForUpgrade.List()
+}
+
+// getLatestNextVersion returns the next available version for a provider within the target ClusterAPI/API version.
+// the next available version is tha latest version available in the for the target ClusterAPI/API version.
+func (i *upgradeInfo) getLatestNextVersion(targetAPIVersion string) *version.Version {
+	var latestNextVersion *version.Version
+	for _, releaseSeries := range i.metadata.ReleaseSeries {
+		// Skip the release series if not linked with the target ClusterAPI/API version version
+		if releaseSeries.ClusterAPIVersion != targetAPIVersion {
+			continue
+		}
+
+		for j := range i.nextVersions {
+			nextVersion := &i.nextVersions[j]
+
+			// Drop the nextVersion version if not linked with the current release series
+			if nextVersion.Major() != releaseSeries.Major || nextVersion.Minor() != releaseSeries.Minor {
+				continue
+			}
+
+			// Drop the nextVersion if older that the latestNextVersion selected so far
+			if latestNextVersion == nil || latestNextVersion.LessThan(nextVersion) {
+				latestNextVersion = nextVersion
+			}
+		}
+	}
+
+	return latestNextVersion
+}
+
+// versionTag converts a version to a RepositoryTag
+func versionTag(version *version.Version) string {
```

Comment:
Minor: This version is used within the `githubRepostiry` and passed in as tag. That tag doesn't get checked and gets passed into `Repositories.GetReleaseByTag` which also doesn't check if the version is empty but assumes the tag will exist.

Does it make sense to provide a better default here or even return an error if version is nil? Otherwise this will error when trying to get Metadata and then fall into the case of using the embed metadata. 

Another suggestion would be to have better error handling in the Metadata Get function. Right now, it assumes any error means it should just use the embedded metadata, but my understanding is that a 404 would want to use embedded metadata (since that would imply metadata doesn't exist for a repository) and every other error would actually be an error. If we don't add this error check then the statement I found in some comment that the provider metadata always overrules embedded metadata could end up not being true if, say, GitHub is having a bad day.

Reply:
`versionTag` is used in two points in the code, and in both the version is a result of a `version.ParseSemantic` with error check.

Also in the `GitHubRepository` we are considering only tags which are semantic versions, so this should never be nil.